riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+            let (rd, r1, r2) = args.rrr()?;
+            only_if_no_write_to_zero_vec(
+                rd,
+                vec![
+                    //format!("to_signed {}, {};", r1.addr(), tmp1.addr()),
+                    format!("affine {}, {}, 0, 1, 0, 0;", r1.addr(), tmp1.addr()),
+                    // tmp2 is 1 if tmp1 is non-negative
+                    format!(
+                        "is_greater_or_equal_signed {}, 0, {};",
+                        tmp1.addr(),
+                        tmp2.addr()
+                    ),
+                    // If negative, convert to positive
+                    format!("skip_if_equal 0, {}, 0, 1, 1;", tmp2.addr()),
+                    format!(
+                        "affine {}, {}, {}, {}, 0, 0;",
+                        tmp1.addr(),
+                        tmp1.addr(),
+                        i32_high(-1),
+                        i32_low(-1)
+                    ),
+                    format!(
+                        "mul {}, {}, {}, {};",
+                        tmp1.addr(),
+                        r2.addr(),
+                        tmp1.addr(),
+                        rd.addr()
+                    ),
+                    // If was negative before, convert back to negative
+                    format!("skip_if_equal 0, {}, 0, 1, 2;", tmp2.addr()),
+                    format!("is_equal_zero {}, {};", tmp1.addr(), tmp1.addr()),
+                    // If the lower bits are zero, return the two's complement,
+                    // otherwise return one's complement.
+                    format!(
+                        "sub_wrap_with_offset {}, {}, {}, {}, {};",
+                        tmp1.addr(),
+                        rd.addr(),
+                        i32_high(-1),
+                        i32_low(-1),
+                        rd.addr()
+                    ),
+                ],
+            )
+        }
+        "divu" => {
+            let (rd, r1, r2) = args.rrr()?;
+            only_if_no_write_to_zero(
+                rd,
+                format!(
+                    "divremu {}, {}, {}, {};",
+                    r1.addr(),
+                    r2.addr(),
+                    rd.addr(),
+                    tmp1.addr()
+                ),
+            )
+        }
+        "remu" => {
+            let (rd, r1, r2) = args.rrr()?;
+            only_if_no_write_to_zero(
+                rd,
+                format!(
+                    "divremu {}, {}, {}, {};",
+                    r1.addr(),
+                    r2.addr(),
+                    tmp1.addr(),
+                    rd.addr()
+                ),
+            )
+        }
+
+        // bitwise
+        "xor" => {
+            let (rd, r1, r2) = args.rrr()?;
+            only_if_no_write_to_zero(
+                rd,
+                format!("xor {}, {}, 0, 0, {};", r1.addr(), r2.addr(), rd.addr()),
+            )
+        }
+        "xori" => {
+            let (rd, r1, imm) = args.rri()?;
```

Comment:
Idea: if `rri` returned `(Register, Register, U32)` where U32 in a wrapper around U32 whose Display is implemented as `"{}, {}", u32_high(self.0), u32_low(self.0)` then there would be no diff here vs large fields. Not required for this PR though.